agent/config/benchmark/uperf.json

```diff
@@ -0,0 +1,71 @@
+{
+  "controller" : {
+    "pre-script" : "/opt/pbench-agent/bench-scripts/postprocess/uperf-prepare-xmlfile",
+    "post-script" : "/opt/pbench-agent/bench-scripts/postprocess/uperf-postprocess-cdm",
+    "parameters" : { 
+      "defaults": {
+        "basic" : [ "--test-type=stream", "--protocol=tcp", "--message-size=16384",
+          "--instances=1", "--runtime=30"
+          ],
+        "full" : [ "--test-type=stream,maerts", "--protocol=tcp,udp",
+          "--message-size=64,256,1024,4096,16384,32768", "--instances=1,8",
+          "--runtime=30"
+          ],
+        "mandatory" : []
+      },
+      "specs" : {
+        "size_KMG" : {
+          "description" : "bytes in k/K (1024), m/M (1024^2), g/G (1024^3) or no units: 256 4k 16M 1g",
+          "arguments" : [ "message-size", "wsize", "rsize" ],
+       	  "value_regex" : "[0-9]+[kbmgKBZZMG]{0,1}",
+       	  "value_transforms" : [ "s/([0-9]+)[gG]/($1*1024).\"M\"/e",
+            "s/([0-9]+)[mM]/($1*1024).\"K\"/e", "s/([0-9]+)[kK]/($1*1024)/e"
+            ]
+          },
+        "hosts-or-IPs" : {
+          "description" : "list of valid hostnames or IPs",
+          "arguments" : [ "remotehost" ],
+          "value_regex" : "\\S+"
+          },
+        "test-types" : {
+          "description" : "all possible uperf test types",
+          "arguments" : [ "test-type" ],
+          "value_regex" : "^stream$|^maerts$|^rr$"
+          },
+        "protocols" : {
+          "description" : "all possible protocols",
+          "arguments" : [ "protocol" ],
+          "value_regex" : "^tcp|udp$"
+          },
+        "time_smh" : {
+          "description" : "time in seconds, minutes, or hours: 10s 2m 1h",
+          "arguments" : [ "runtime" ],
+          "value_regex" : "^[0-9]+[smh]{0,1}$",
+       	  "value_transforms" : [ "s/([0-9]+)[hH]/($1*60).\"M\"/e",
+            "s/([0-9]+)[mM]/($1*60)/e"
+            ]
+          },
+        "positive_integer" : {
+          "description" : "a whole number greater than 0",
+          "arguments" : [ "instances" ],
+          "value_regex" : "[1-9][0-9]*"
+          }
+        }
+      }
+    },
+  "client" : { 
+    "packages" : [ "pbench-uperf-1.0.4-1" ],
+    "pre-script" : " ",
+    "post-script" : " ",
+    "copy-to" : [ "/var/lib/pbench-agent/tmp/uperf.xml" ],
```

Comment:
Why do we need to encode the `copy-to` location to be a temporary directory?  This seems like an execution time value that shouldn't be in the JSON, but not sure about it.

Reply:
I may be speaking for Andrew here, but I think the point is to move all benchmark specific logic like this into the JSON configuration file so that the generic scripts (pbench-run-benchmark, pbench-gen-iterations, pbench-run-benchmark-sample) can just use these well know field names to access the necessary information to perform well known steps like copying a config file to a client system.

There is very little (if any) execution time scripting that would be benchmark specific and be able to handle these types of things for us.  The major exception to that is postprocessing which obviously tends to be very specific to the various output(s) that a benchmark produces.

Reply:
The copy-to: is actually specifying the *source* of the files that need to be copied, which could be from anywhere.  See (https://github.com/distributed-system-analysis/pbench/issues/1320) for not hardcoding things like /var/lib/pbench-agent/tmp/ and using a pbench_var.  

The destination for copy-to is fixed and the remote sample directory for each client.

Reply:
Can you explain more?  If the destination is fixed, why does it need to be in the JSON specification?  What entity consumes the value of `copy-to`?

Reply:
The destination is fixed, but the names of the files to "copy-to" the clients/servers are not.  This is where a benchmark specifies those files.  Each benchmark may want different files copied from the agent to the clients/servers (fio.job for fio, uperf.xml for uperf, etc)  It is up to pbench-run-benchmark-sample to provide this functionality to the benchmark, as long as the benchmark json provides the source files it needs.  Please see https://github.com/distributed-system-analysis/pbench/blob/5f9773d83e05968e64d93aec7ea14385fde07f1b/agent/bench-scripts/pbench-run-benchmark-sample#L123